THE ASSIGNMENT IS ALL ABOUT LOADING THE DATASET AND PERFORM EDA PROCESS ON TOP OF THE DATASET WHICH IS IN PARQUET FILE FORMAT.


In [1]:
%pip install numpy==1.23.5
%pip install pandas==1.5.3
%pip install sweetviz==2.1
%pip install pyarrow

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd 
import numpy as np
import os
# import pyarrow

Loading data as parquet files

In [3]:
source_path="/home/ai/Documents/my files/assignment_2/Pyspark_Assignment_2/hotelsourcesdetails.parquet"

In [4]:
hotel_details=pd.read_parquet(source_path)
hotel_details.head(5)

id  hotelcode   hotelid  websitecode                    hotelname  \
0  4492402     329109  22995521            1                Hotel Perales   
1  4492405     329778   1381371            1  The Cherry Tree Guest House   
2  4492408     333046  17810920            1            Asia Boutique Inn   
3  4492409     334152  46258460            1          Dunham's Bay Resort   
4  4492410     334788  31836214            1          Pazo de Turbisquedo   

                                             address                  city  \
0                                 Avenida Pio XII, 3  Talavera de la Reina   
1                                 15 Silverdale Road            Eastbourne   
2  Rua Correia Da Silva No. 146 e terreno junto a...                 Macao   
3                                2999 State Route 9L           Lake George   
4                           Lugar de Turbisquedo s/n           Turbisquedo   

                state countrycode       zip        lat         lng  status  \
0  Castilla-La Mancha         ESP     45600  39.963929   -4.819594       1   
1         East Sussex         GBR  BN20 7AJ  50.760040    0.281270       1   
2               Hunan         CHN      None  22.152700  113.557000       1   
3                  NY         USA     12845  43.441875  -73.649711       1   
4             Galicia         ESP     32151  42.482576   -7.763914       1   

      propertytype  rating  larbarterm  reviewscount  
0            Hotel     3.0         NaN           NaN  
1      Guest house     4.0         NaN           NaN  
2              Inn     2.0         NaN           NaN  
3            Hotel     3.0         NaN           NaN  
4  Bed & Breakfast     3.0         NaN           NaN

In [5]:
irl_hotel_details=hotel_details[hotel_details["countrycode"]=="IRL"]
irl_hotel_details.head(5)

id  hotelcode   hotelid  websitecode                   hotelname  \
155  4492621     459664  15690635            1  Auburndale Bed & Breakfast   
570  4493300     959939  46310864            1     Hillside Haven cottages   
773  4493646    1212015  26991076            1          Ocean View Doogort   
834  4493805    1294694  40029197            1             Rose of the Sea   
910  4493931    1350340    176109            1                    The Mont   

                         address          city             state countrycode  \
155  Smithland East, Springmount      Kilkenny   County Kilkenny         IRL   
570        82 1, Jayawardanapura  Nuwara Eliya  Central Province         IRL   
773            Ocean view Dugort       Doogort          Connacht         IRL   
834                Corclogh East     Belmullet          Connacht         IRL   
910               Merrion Square        Dublin          Leinster         IRL   

      zip        lat        lng  status  propertytype  rating  larbarterm  \
155  None  52.634443  -7.251345       1   Guest house     4.0         NaN   
570  None   6.923745  80.792331       1  Holiday home     3.0         NaN   
773  None  54.013329 -10.017120       0  Holiday Home     3.0         NaN   
834  None  54.205300  -9.963980       1      Homestay     0.0         NaN   
910  None  53.341477  -6.250687       1         Hotel     4.0         NaN   

     reviewscount  
155           NaN  
570           NaN  
773           NaN  
834           NaN  
910           NaN

In [6]:
irl_hotel_details.isnull().mean()*100

id                0.000000
hotelcode         0.000000
hotelid           0.000000
websitecode       0.000000
hotelname         0.000000
address           0.107421
city              0.007408
state             8.652974
countrycode       0.000000
zip              97.616358
lat               0.000000
lng               0.000000
status            0.000000
propertytype      0.000000
rating            8.630748
larbarterm      100.000000
reviewscount     92.662012
dtype: float64

In [7]:
hotel_rates=pd.read_parquet("/home/ai/Documents/my files/assignment_2/Pyspark_Assignment_2/hospitality_raw_rates_20240922_123949.parquet")
hotel_rates.head(5)

websitecode         dtcollected  los  guests               roomtype  \
0            2 2024-09-22 18:19:47    1       1                   None   
1            2 2024-09-22 18:19:47    1       1  One-Bedroom Apartment   
2            2 2024-09-22 18:19:47    1       1  One-Bedroom Apartment   
3            2 2024-09-22 18:19:47    1       1  One-Bedroom Apartment   
4            2 2024-09-22 18:19:47    1       1  One-Bedroom Apartment   

   onsiterate  netrate currency  \
0         NaN      NaN      USD   
1        70.0     78.0      USD   
2        63.0      NaN      USD   
3        70.0      NaN      USD   
4        70.0      NaN      USD   

                                     roomdescription  \
0                                                nan   
1  Bedroom 1: 1 full bed , Living room: 1 sofa be...   
2  Bedroom 1: 1 bunk bed , Living room: 1 sofa be...   
3  Bedroom 1: 1 bunk bed , Living room: 1 sofa be...   
4  Bedroom 1: 1 bunk bed , Living room: 1 sofa be...   

                                    rateplan  ... pos  \
0                                        nan  ...   0   
1   Free cancellation before October 8, 2024  ...   0   
2                                          •  ...   0   
3  Free cancellation before October 12, 2024  ...   0   
4                                          •  ...   0   

                                           proxyused  israteperstay mealplan  \
0  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...            nan      nan   
1  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...              Y      nan   
2  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...              Y      nan   
3  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...              Y      nan   
4  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...              Y      nan   

  optionalextras payplan id    refid   hotelid  minstay  
0            NaN     NaN  1   262438   4157796      NaN  
1            NaN     NaN  2  2055755   6578136      1.0  
2            NaN     NaN  3  1972651  10804142      1.0  
3            NaN     NaN  4  1972651  10804142      1.0  
4            NaN     NaN  5  1972651  10804142      1.0  

[5 rows x 33 columns]

In [8]:
unique_hotels = irl_hotel_details['hotelname'].unique()

# Print the number of unique hotels
print("Number of unique hotels:", len(unique_hotels))

Number of unique hotels: 16447


CHOOSING OUT ID FIELD TO JOIN TWO DATA FRAMES

In [9]:
print(irl_hotel_details['id'].nunique(), hotel_rates['id'].nunique())
print(irl_hotel_details['hotelid'].nunique(), hotel_rates['hotelid'].nunique())
print(irl_hotel_details['websitecode'].nunique(), hotel_rates['websitecode'].nunique())

53993 8297401
38648 2324483
31 1


id

irl_hotel_details: 53993 unique values
hotel_rates: 8297401 unique values
The id column in hotel_rates has a much larger number of unique values than in irl_hotel_details. This suggests that id might not be a suitable join key, as it may not uniquely identify hotels in both tables.

hotelid

irl_hotel_details: 38648 unique values
hotel_rates: 2324483 unique values
Similar to the id column, the hotelid column in hotel_rates has a larger number of unique values than in irl_hotel_details. However, the difference is not as drastic as with the id column.

websitecode

irl_hotel_details: 31 unique values
hotel_rates: 1 unique value
The websitecode column in hotel_rates has only 1 unique value, which is likely not suitable for joining the tables.

based on this I find that hotel is good to choose for join two data frames


In [10]:
import sweetviz as sv
from PIL import Image


# Create a Sweetviz report
report = sv.analyze(irl_hotel_details)

# Show the report
report.show_html('ireland_df_report.html')

/home/ai/.local/lib/python3.10/site-packages/sweetviz/dataframe_report.py:75: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  all_source_names = [cur_name for cur_name, cur_series in source_df.iteritems()]
/home/ai/.local/lib/python3.10/site-packages/sweetviz/dataframe_report.py:110: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  filtered_series_names_in_source = [cur_name for cur_name, cur_series in source_df.iteritems()


                                             |          | [  0%]   00:00 -> (? left)

/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the same result, you may do `(df - df.mean()).abs().mean()`.
  stats["mad"] = series.mad()
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the same result, you may do `(df - df.mean()).abs().mean()`.
  stats["mad"] = series.mad()
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_text.py:19: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for item in to_process.source_counts["value_counts_without_nan"].iteritems():
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the sam

Report ireland_df_report.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


sweetviz for hotel rates dataframes

In [11]:
# import sweetviz as sv
# from PIL import Image


# # Create a Sweetviz report
# report = sv.analyze(hotel_rates)

# # Show the report
# report.show_html('ireland_df_report.html')

COMPARING TWO DATAFRAMES

In [12]:
comparision_report=sv.compare([irl_hotel_details,"Hotel Details"],[hotel_rates,"Hotel Rates"])
# show_logo=0?
comparision_report.show_html("hotel_comparison_report.html")

/home/ai/.local/lib/python3.10/site-packages/sweetviz/dataframe_report.py:75: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  all_source_names = [cur_name for cur_name, cur_series in source_df.iteritems()]
/home/ai/.local/lib/python3.10/site-packages/sweetviz/dataframe_report.py:93: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  all_compare_names = [cur_name for cur_name, cur_series in compare_df.iteritems()]
/home/ai/.local/lib/python3.10/site-packages/sweetviz/dataframe_report.py:110: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  filtered_series_names_in_source = [cur_name for cur_name, cur_series in source_df.iteritems()


                                             |          | [  0%]   00:00 -> (? left)

/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the same result, you may do `(df - df.mean()).abs().mean()`.
  stats["mad"] = series.mad()
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the same result, you may do `(df - df.mean()).abs().mean()`.
  stats["mad"] = series.mad()
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the same result, you may do `(df - df.mean()).abs().mean()`.
  stats["mad"] = series.mad()
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_text.py:19: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items inste

Report hotel_comparison_report.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


CREATE A DICTIONARY BY DESCRIBING EACH FIELD IN HOTEL DETAILS DATAFRAME

In [13]:
field_descriptions1 = {
  "id": "A unique identifier for each hotel record.",
  "hotelcode": "A code or identifier specific to the hotel.",
  "hotelid": "A unique identifier for the hotel, possibly used across multiple websites or systems.",
  "websitecode": "A code or identifier for the website or platform where the hotel information is listed.",
  "hotelname": "The name of the hotel.",
  "address": "The street address of the hotel.",
  "city": "The city where the hotel is located.",
  "state": "The state or province where the hotel is located.",
  "countrycode": "A code representing the country where the hotel is located.",
  "zip": "The postal code or zip code of the hotel's location.",
  "lat": "The latitude coordinate of the hotel's location.",
  "lng": "The longitude coordinate of the hotel's location.",
  "status": "The current status of the hotel (e.g. active, inactive, etc.).",
  "propertytype": "The type of property (e.g. hotel, resort, etc.).",
  "rating": "The hotel's rating or score, possibly based on reviews or other criteria.",
  "larbarterm": "Unknown",
  "reviewscount": "The number of reviews or ratings the hotel has received."
}
print(field_descriptions1)

{'id': 'A unique identifier for each hotel record.', 'hotelcode': 'A code or identifier specific to the hotel.', 'hotelid': 'A unique identifier for the hotel, possibly used across multiple websites or systems.', 'websitecode': 'A code or identifier for the website or platform where the hotel information is listed.', 'hotelname': 'The name of the hotel.', 'address': 'The street address of the hotel.', 'city': 'The city where the hotel is located.', 'state': 'The state or province where the hotel is located.', 'countrycode': 'A code representing the country where the hotel is located.', 'zip': "The postal code or zip code of the hotel's location.", 'lat': "The latitude coordinate of the hotel's location.", 'lng': "The longitude coordinate of the hotel's location.", 'status': 'The current status of the hotel (e.g. active, inactive, etc.).', 'propertytype': 'The type of property (e.g. hotel, resort, etc.).', 'rating': "The hotel's rating or score, possibly based on reviews or other criter

In [14]:
field_descriptions2={
    "websitecode": "Unique code for the website",
    "dtcollected": "Date and time of data collection",
    "los": "Length of stay",
    "guests": "Number of guests",
    "roomtype": "Type of room",
    "onsiterate": "On-site rate",
    "netrate": "Net rate",
    "currency": "Currency of the rate",
    "roomdescription": "Description of the room",
    "rateplan": "Rate plan",
    "pos": "Point of sale",
    "proxyused": "Proxy used for booking",
    "israteperstay": "Indicates if rate is per stay",
    "mealplan": "Meal plan",
    "optionalextras": "Optional extras",
    "payplan": "Payment plan",
    "id": "Unique identifier",
    "refid": "Reference ID",
    "hotelid": "Hotel ID",
    "minstay": "Minimum stay required"
}
print(field_descriptions2)

{'websitecode': 'Unique code for the website', 'dtcollected': 'Date and time of data collection', 'los': 'Length of stay', 'guests': 'Number of guests', 'roomtype': 'Type of room', 'onsiterate': 'On-site rate', 'netrate': 'Net rate', 'currency': 'Currency of the rate', 'roomdescription': 'Description of the room', 'rateplan': 'Rate plan', 'pos': 'Point of sale', 'proxyused': 'Proxy used for booking', 'israteperstay': 'Indicates if rate is per stay', 'mealplan': 'Meal plan', 'optionalextras': 'Optional extras', 'payplan': 'Payment plan', 'id': 'Unique identifier', 'refid': 'Reference ID', 'hotelid': 'Hotel ID', 'minstay': 'Minimum stay required'}


USING SWEETVIZ LIBRARY TO ANALYZE AND MAKE REPORTS


In [15]:
# import sweetviz as sv
# from PIL import Image


# # Create a Sweetviz report
# report = sv.analyze(ireland_df)

# # Show the report
# report.show_html('ireland_df_report.html')


DESCRIBING EACH FIELD WITH ITS PREPROCESSING AND STANDARDIZATION PROCESS.


In [16]:
import csv

data = [
    ["Field", "Description", "Preprocessing", "Standardization"],
    ["ID", "This is unique variable in both the dataframes", "Not required", "Already consists unique values"],
    ["hotelcode", "It contains only 8470 unique values which is about 27%", "Not required", "Already consists unique values"],
    ["hotelid", "this column basically contains the id of the hotel", "Not required", "Already consists unique values"],
    ["websitecode", "the website code column has no null values and it is highly correlated with status and propertytype", "Not required", "Already consists unique values"],
    ["hotel name", "this column especially talks about the name of the hotel in the ireland", "convert the name into lower case. removal of special character", "normalizing the hotel name into standard format"],
    ["Address", "this address column states the where the particular hotel is located", "Handle the missing values may be we can impute the values since there is only 49. else we can replace the null values with 'unknown value'", "normalizing it to the standard format"],
    ["city", "In this column the it contains only 1336 unique values where the most of the hotels located in one or more cities", "Handling the missing values by imputing", "normalizing it to the standard format"],
    ["state", "In the state column we have about 98% records", "There is no missing values and preprocessing includes removal of punctuation marks", "normalize to standard format"],
    ["countrycode", "to talk about country code we have filtered only one country that is ireland", "not required", "not required"],
    ["zip code", "It is a feature used in foriegn countries to deliver the mail faster", "Since there is a lot of missing values we could drop the column else we could map with the cities and impute the null values accordingly", "There is particular format for zip code according to their cities"],
    ["latitude", "this latitude region will show only the latitude of the region in ireland", "not required", "Already consists unique values"],
    ["longitude", "this is similar to latitude but it contains 8521 unique values", "not required", "Already consists unique values"],
    ["status", "this columns shows whether these hotels are active or not by indicating 1 to show it is active and 0 to show it is inactive", "not required", "It is already standardized"],
    ["propertytype", "this column tell about wether the property is a hotel or apartments or hostel or guest house and so on", "There is no missing values and removal of punction marks are necessary", "Should be with in a range"],
    ["rating", "the rating column has only 10 uniques values which ranges from 1-5 with the interval of 0.5", "Here handling missing values is required, but then we cannot give rating based on mode that would be inappropriate", "not required"],
    ["larberterm", "This column has no values so, it is difficult to know what actually larbarterm refers to", "This field is not required better you can drop the column", "not required"],
    ["reviewscount", "This column basically means the counts of the review and it also contains 99% of the null values", "Better to drop the column", "not required"],
    ["website rates", "which is basically contains only 1 value due to the filteration of ireland data", "not required", "not required"],
    ["dtcollected", "This might be the dates that is the data collected on", "Convert to a standard date format (e.g., YYYY-MM-DD)", "not required"],
    ["los", "this value is based on the fileteration may be", "not required", "not required"],
    ["guests", "this value is based on the fileteration may be", "not required", "not required"],
    ["room type", "This is column talks about the type of rooms that the hotels have", "Handle the missing values by imputing", "Covert to the standard format"],
    ["on site rate", "this speaks about the rate of the site whcih have 1140 distinct values and 4803 missing values", "The missing values should be handled by imputing the values", "Can be normalized to a standard scale(eg.1-100)"],
    ["net rate", "net rate has a huge impact on the hotels growth", "There is a huge null values. It cannot be imputed as well", "Can be normalize to a standard scale(eg 0-100)"],
    ["currency", "currency is basicaly here represents the currency used in ireland", "not required", "not required"],
    ["room amenities","This is clear that it talks about the facilities in the hotels or apartments or any kinda property","Handle the missing values by imputing but however it wouldn't be fine. The punctuation and stop words has to be removed","not required"],
    ["occupany","this column is about how many people can be accomodated within that property","Not Required","wihtin a scale of 1-4"],
    ["ispromo","I guess it might talk about the promocode for the discounts","not required","not required"],
    ["closed","This columns talks about whether the property is open or closed","Not required","not required"],
    ["checkin","This data is collected on the date 13.10.2024 so there is only one value in this column","Not Required","not required"],
    ["checkout","This datais also collected on the date 14.10.2024","Not Required","not required"],
    ["discount","It speaks about whether there is any discount for the people who are booking on the particular date","need to handle the null value by providing it as 0","not required"],
    ["promo name","It tells about the promo code like new propert deal , last escape deal and so on","These promo name is a unique contraints and the missing values has to imputed by maping with discount or amenities","Convert it into a standard format"],
    ["status code","It usually speaks about whether the promo are active","not required","not required"],
    ["taxstatus","It is all about tax for th no.of people staying in the hotels","not required","Should  set a scale limit"],
    ["taxtype","amount plus, taxes and fee  for staying only fee and taxes","While encoding we could make a label encoder.should handle missing values","not required"],
    ["taxamount","the tax amount is mostly between 10-20 and most of the taxes are between the earlier mentioned range and few are between 20 -50 and rarely above 60","better this can be drioped else we could make some calculation and fix it","not required"],
    ["pos","I don't understand about this pos might be it talks about the position","Not Required","Not Required"],
    ["proxy used","There are nearly 100 unique proxy used to get the data from the websites","confused","not required"],
    ["is rate per stay","This column shows that whether there is rate included per stay","The missing values has to be handled","not required"],
    ["meal plan","This is important feature to note because if there is anything like no meal provided poeple need to search for the restaurant to have their meal","IT has to be preprocessed in such a way that if breakfast is provided it should be encoded as 1 .. likewise for other","not required if it is encode properly"],
    ["optional extras","This has NO value and it is difficult to find what is signifies","It is better to drop the column","not required"],
    ["payplan","this also the same. there is no plan given for any of the records","Good to drop the column","not required"],
    ["ref id","ref id signifies that there is any referal id for a paricular stay or property","may be this could be mapped with the promo code or discount and make some calculations out of it","not required"],
    ["hotelid rates","this column has a large no seems to be the rate of the hotel id","confused","not required"],
    ["min stay","this column has only one value as 1. most probably seems to be minimum no.of people can stay","missing values has to be handled or replace with ones","not required"]
]

with open('Analysis.csv', 'w', newline='') as file:
    writer = csv.writer(file)
    writer.writerows(data)

In [17]:
analysis_df=pd.read_csv("/home/ai/Documents/my files/assignment_2/Pyspark_Assignment_2/Analysis.csv")
analysis_df.head(5)

Field                                        Description  \
0           ID     This is unique variable in both the dataframes   
1    hotelcode  It contains only 8470 unique values which is a...   
2      hotelid  this column basically contains the id of the h...   
3  websitecode  the website code column has no null values and...   
4   hotel name  this column especially talks about the name of...   

                                       Preprocessing  \
0                                       Not required   
1                                       Not required   
2                                       Not required   
3                                       Not required   
4  convert the name into lower case. removal of s...   

                                   Standardization  
0                   Already consists unique values  
1                   Already consists unique values  
2                   Already consists unique values  
3                   Already consists unique values  
4  normalizing the hotel name into standard format

In [18]:
hotels_id_cols = ['id', 'hotelcode', 'hotelid']
rates_id_cols = ['id', 'refid', 'hotelid']
for h_id in hotels_id_cols:
    for r_id in rates_id_cols:
        num_r_h = hotel_rates[r_id].nunique()
        common = len( set(irl_hotel_details[h_id]) & set(hotel_rates[r_id]) )
        print('hotel column used:', h_id, '. rate column used:', r_id, '. Pct common:', (common / num_r_h) )
        print('num_r_h:', num_r_h)

hotel column used: id . rate column used: id . Pct common: 0.0037559954014516114
num_r_h: 8297401
hotel column used: id . rate column used: refid . Pct common: 0.003750061841389205
num_r_h: 2324495
hotel column used: id . rate column used: hotelid . Pct common: 0.0
num_r_h: 2324483
hotel column used: hotelcode . rate column used: id . Pct common: 0.0020794463230112657
num_r_h: 8297401
hotel column used: hotelcode . rate column used: refid . Pct common: 0.002551952144444277
num_r_h: 2324495
hotel column used: hotelcode . rate column used: hotelid . Pct common: 0.0
num_r_h: 2324483
hotel column used: hotelid . rate column used: id . Pct common: 0.0
num_r_h: 8297401
hotel column used: hotelid . rate column used: refid . Pct common: 0.0
num_r_h: 2324495
hotel column used: hotelid . rate column used: hotelid . Pct common: 0.003527235948810983
num_r_h: 2324483
